packages/opentelemetry-plugin-fetch/src/fetch.ts

```diff
@@ -0,0 +1,351 @@
+      const plugin = this;
+
+      return function patchConstructor(
+        this: (input: RequestInfo, init?: RequestInit) => Promise<Response>,
+        input: RequestInfo,
+        init?: RequestInit
+      ): Promise<Response> {
+        const url = input instanceof Request ? input.url : input;
+        const options: RequestInit =
+          input instanceof Request ? input : init || {};
+
+        const span = plugin._createSpan(url, options);
+        if (!span) {
+          return original.apply(this, [url, options]);
+        }
+        const spanData = plugin._prepareSpanData(url);
+
+        function onSuccess(
+          span: api.Span,
+          resolve: (
+            value?: Response | PromiseLike<Response> | undefined
+          ) => void,
+          response: Response
+        ) {
+          try {
+            if (response.status >= 200 && response.status < 400) {
+              plugin._endSpan(span, spanData, response);
+            } else {
+              plugin._endSpan(span, spanData, {
+                status: response.status,
+                statusText: response.statusText,
+                url,
+              });
+            }
+          } finally {
+            resolve(response);
+          }
+        }
+
+        function onError(
+          span: api.Span,
+          reject: (reason?: unknown) => void,
+          error: FetchError
+        ) {
+          try {
+            plugin._endSpan(span, spanData, {
+              status: error.status || 0,
+              statusText: error.message,
+              url,
+            });
+          } finally {
+            reject(error);
+          }
+        }
+
+        return new Promise((resolve, reject) => {
+          return plugin._tracer.withSpan(span, () => {
+            plugin._addHeaders(options, url);
+            plugin._tasksCount++;
+            return original
+              .apply(this, [url, options])
+              .then(
+                onSuccess.bind(this, span, resolve),
+                onError.bind(this, span, reject)
+              );
+          });
+        });
+      };
+    };
+  }
+
+  /**
+   * Prepares a span data - needed later for matching appropriate network
+   *     resources
+   * @param spanUrl
+   */
+  private _prepareSpanData(spanUrl: string): SpanData {
+    const startTime = core.hrTime();
+    const entries: PerformanceResourceTiming[] = [];
+    const observer: PerformanceObserver = new PerformanceObserver(list => {
+      const entries = list.getEntries() as PerformanceResourceTiming[];
+      entries.forEach(entry => {
+        if (entry.initiatorType === 'fetch' && entry.name === spanUrl) {
+          entries.push(entry);
+        }
+      });
+    });
+    observer.observe({
+      entryTypes: ['resource'],
+    });
+    return { entries, observer, startTime, spanUrl };
+  }
+
+  /**
+   * implements patch function
+   */
+  patch() {
+    if (core.isWrapped(window.fetch)) {
```

Comment:
What level of compatibility do you see for this plugin? `fetch` seems to be the "universal" JS API for http requests and so it is available on non-DOM / non-browser environments as well, e.g. react-native. It seems like checking for `global`, `window`, `self`, etc everywhere in this plugin can make this work almost everywhere. There are some instances where this isn't the case, like I see this plugin uses `addEventListener`. If you think it is easier to only target web for now that is fine, but I think the only thing missing is adding a bunch of `typeof window !== 'undefined'` checks.

Reply:
It is targetted to browser only now. If some modification can be made to support it wider, then I guess I would do it in separate PR with appropriate tests.